### PR TITLE
feat: update password service

### PR DESCRIPTION
**Pull request by oopshell**
#### Added resolver of changeing password:
- get ID of the current logged in user;
- verify current password;
- replace old password with new one
- add password pattern validation pipe

#### Added unit test:
- add test for changePassword service
- add test for password pattern validation pipe

#### Resolve CP-34

Screenshot 2024-04-17 at 03 30 28
Screenshot 2024-04-17 at 03 31 01
Screenshot 2024-05-29 at 17 46 10

### CODE REVIEW

File: src/modules/auth/auth.resolver.ts [hunk 1]
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
       message: 'registration failed',
     };
   }
+
+  @Mutation(() => Result, { description: 'Change password' })
+  @UseGuards(GqlAuthGuard)
+  async changePassword(
+    @Context() cxt: any,
```
Comment (Chris-Wang) [inline src/modules/auth/auth.resolver.ts]: avoid `any`, add context type

File: src/modules/auth/auth.resolver.ts [hunk 2]
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
...
+    const user = await this.userService.find(id);
+    if (!user) {
+      return {
+        code: ACCOUNT_NOT_EXIST,
+        message: "account doesn't exist",
+      };
+    }
+    const isPasswordValid = await bcrypt.compare(oldPassword, user.password);
+    if (isPasswordValid) {
+      const hashedPassword = await bcrypt.hash(newPassword, 10);
```
Comment (Chris-Wang) [inline src/modules/auth/auth.resolver.ts]: consider add new password validation

File: src/modules/auth/auth.resolver.ts [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
-import { Args, Mutation, Resolver } from '@nestjs/graphql';
+import { Args, Context, Mutation, Resolver } from '@nestjs/graphql';
 import { UserService } from '../user/user.service';
+import { UseGuards } from '@nestjs/common';
+import { GqlAuthGuard } from '@/common/guards/auth.guard';
 import { CreateUserInput } from '../user/dto/new-user.input';
 import {
   ACCOUNT_EXIST,
   ACCOUNT_NOT_EXIST,
+  UPDATE_ERROR,
```
Comment (Chris-Wang) [inline src/modules/auth/auth.resolver.ts]: to be more specific, I think we need to add `UPDATE_PASSWORD_ERROR`

File: src/modules/auth/auth.resolver.ts [hunk 4]
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
...
     };
   }
+
+  @Mutation(() => Result, { description: 'Change password' })
+  @UseGuards(GqlAuthGuard)
+  async changePassword(
+    @Context() cxt: any,
+    @Args('oldPassword') oldPassword: string,
+    @Args('newPassword') newPassword: string,
+  ): Promise<Result> {
```
Comment (Chris-Wang) [inline src/modules/auth/auth.resolver.ts:40]: let's add try-catch block in the resolver to catch any unexpected error, and return the error text in response

**Review by Chris-Wang — CHANGES_REQUESTED** [review]
Please add unit test

**Comment by oopshell** [comment]
> Please add unit test

Hi @Chris-Wang , I used **reset --hard** to discard some clumsy commits, and now one of your review comments becomes unresolvable. Could you kindly re-review the PR? I apologise for the trouble and appreciate your help. Thank you!

File: src/modules/auth/auth.resolver.ts [hunk 5]
```diff
@@ -26,4 +27,14 @@ export class AuthResolver {
   async register(@Args('input') input: CreateUserInput): Promise<Result> {
     return await this.authService.register(input);
   }
+
+  @Mutation(() => Result, { description: 'Change password' })
+  @UseGuards(GqlAuthGuard)
+  async changePassword(
+    @Context() cxt: { req: Partial<Request> & { user: {id: string} } },
```
Comment (Chris-Wang) [inline src/modules/auth/auth.resolver.ts]: `{ user: { id: string } }` leave space after or before braces

File: src/modules/auth/auth.service.ts [hunk 6]
```diff
@@ -69,4 +71,46 @@ export class AuthService {
       message: 'registration failed',
     };
   }
+
+  async changePassword(
+    cxt: { req: Partial<Request> & { user: {id: string} } },
```
Comment (Chris-Wang) [inline src/modules/auth/auth.service.ts]: `{ user: { id: string } }` leave space after or before braces

File: src/modules/auth/pipe/password-validation.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import * as Joi from 'joi';
+import { passwordPatten } from '@/common/utils/helpers';
+
+@Injectable()
+export class PasswordValidationPipe implements PipeTransform {
+  private readonly schema = Joi.string().pattern(passwordPatten).messages({
+    'string.pattern.base': 'password must contain 8 to 32 characters, including letter, number and special character.'
+  });
+
+  transform(value: any): string {
```
Comment (Chris-Wang) [inline src/modules/auth/pipe/password-validation.pipe.ts]: consider to use custom exception to give non-error response, 
now we get:
image
here's like what we want:
image

**Review by Chris-Wang — CHANGES_REQUESTED** [review]

File: src/modules/auth/pipe/password-validation.pipe.ts (hunk 7 above)
Comment (oopshell) [reply src/modules/auth/pipe/password-validation.pipe.ts]: Screenshot 2024-05-29 at 17 46 10

**Review by Chris-Wang — APPROVED** [review]
LGTM

**Pull request merged**